pkg/rollout/rollout.go

```diff
@@ -0,0 +1,191 @@
+	svc = r.updateAnnotations(svc, stable, candidate)
+	svc, err = r.RunClient.ReplaceService(serviceName, svc)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not update service")
+	}
+
+	return svc, nil
+}
+
+// SplitTraffic changes the traffic configuration of the service.
+//
+// It creates a new traffic slice in case there's a new candidate revision.
+// If we only drop the traffic of the previous candidate to 0, Cloud Run would
+// still consider it as serving traffic and the slice of traffic targets would
+// grow very large over time.
+func (r *Rollout) SplitTraffic(svc *run.Service, stable, candidate string) *run.Service {
+	candidateTarget, candidatePercent := r.candidateTraffic(svc, candidate)
+	if candidateTarget == nil {
+		candidatePercent = r.Config.Rollout.Steps[0]
+	} else if candidatePercent == candidateTarget.Percent {
+		// If the traffic share did not change, candidate already handled 100%
+		// and is now ready to become stable.
+		r.promoteCandidate = true
+	}
+
+	var traffic []*run.TrafficTarget
+	candidateTarget = newTrafficTarget(candidate, candidatePercent, CandidateTag)
+	if r.promoteCandidate {
+		candidateTarget.Tag = StableTag
+	}
+
+	traffic = append(traffic, candidateTarget)
+	for _, target := range svc.Spec.Traffic {
+		// Respect tags manually introduced by the user.
+		if target.Tag != "" && !target.LatestRevision &&
+			target.Tag != StableTag && target.Tag != CandidateTag {
+
+			traffic = append(traffic, target)
+			continue
+		}
+
+		// When the user introduces a tag manually (UI/gcloud), Cloud Run breaks
+		// all targets with (Name, Tag, Percent) into two different targets
+		// (Name, Tag) and (Name, Percent).
+		// This recovers from this by ignoring the (Name, Tag) target.
+		if (target.Tag == StableTag || target.Tag == CandidateTag) && target.Percent == 0 {
+			continue
+		}
+
+		if target.RevisionName == stable && !r.promoteCandidate {
+			traffic = append(traffic, newTrafficTarget(target.RevisionName, 100-candidatePercent, StableTag))
+		}
+	}
+
+	// Always assign latest tag to the latest revision.
+	traffic = append(traffic, &run.TrafficTarget{LatestRevision: true, Tag: LatestTag})
+
+	if !r.promoteCandidate {
+		r.Log.Printf("Assigning %d%% of the traffic to stable revision %s", 100-candidatePercent, format.Bold(stable))
+		r.Log.Printf("Assigning %d%% of the traffic to candidate revision %s", candidatePercent, format.Bold(candidate))
+	} else {
+		r.Log.Printf("Making revision %s stable\n", format.Bold(candidate))
+	}
+
+	svc.Spec.Traffic = traffic
+
+	return svc
+}
+
+// candidateTraffic returns the traffic configuration for the candidate and the
+// next traffic share for the candidate.
+func (r *Rollout) candidateTraffic(svc *run.Service, candidate string) (*run.TrafficTarget, int64) {
+	for _, target := range svc.Status.Traffic {
+		if target.RevisionName == candidate && target.Percent > 0 {
+			return target, r.nextCandidateTraffic(target.Percent)
```

Comment:
this decision is too deep in the chain, hence not very flexible, move to SplitTraffic